src/transformers/models/paligemma/modeling_paligemma.py

```diff
@@ -306,23 +347,24 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
         )
         final_embedding = torch.where(pad_mask_expanded, torch.zeros_like(final_embedding), final_embedding)
 
-        final_attention_mask_4d = attention_mask.unsqueeze(1).unsqueeze(2) * attention_mask.unsqueeze(1).unsqueeze(-1)
-        final_attention_mask_4d = final_attention_mask_4d.float().expand(
-            -1, self.config.text_config.num_key_value_heads, -1, -1
-        )
-
         # position_ids = torch.arange(0, sequence_length, device=input_ids.device).expand(batch_size, -1)
         # position_ids = torch.where(input_ids == self.pad_token_id, torch.ones_like(position_ids), position_ids)
         position_ids = (attention_mask.cumsum(-1)).masked_fill_((attention_mask == 0), 1)
 
         if labels is not None:
+            causal_mask = self.construct_causal_mask_with_block_attention(
+                attention_mask, labels, text_mask, inputs_embeds
+            )
```

Comment:
Here, the `token_type_ids` can be passed from the forward since `inputs` are unpacked and will contain them. Then, you already have your mask to build the causal mask!